src/main/java/org/opensearch/knn/index/query/rescore/RescoreSearcher.java

```diff
@@ -0,0 +1,101 @@
+     * an oversample factor and then rescoring down to k results for the whole shard.
+     *
+     * @param indexSearcher The index searcher for the current search
+     * @param leafReaderContexts The leaf reader contexts for the current search
+     * @param knnWeight The knn weight for the current search
+     * @param rescoreContext The rescore context for the current search
+     * @return The top k results for the shard
+     * @throws IOException If an error occurs during the search
+     */
+    public TopDocs[] search(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        RescoreContext rescoreContext
+    ) throws IOException {
+        int finalK = ((KNNQuery) knnWeight.getQuery()).getK();
+        int firstPassK = rescoreContext.getFirstPassK(finalK);
+        List<Map<Integer, Float>> perLeafResults = firstPassSearch(indexSearcher, leafReaderContexts, knnWeight, firstPassK);
+        return rescoreResults(indexSearcher, leafReaderContexts, knnWeight, perLeafResults, finalK);
+    }
+
+    private List<Map<Integer, Float>> firstPassSearch(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int k
+    ) throws IOException {
+        List<Callable<Map<Integer, Float>>> tasks = new ArrayList<>(leafReaderContexts.size());
+        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
+            tasks.add(() -> firstPassSearchLeaf(leafReaderContext, knnWeight));
+        }
+        return resultUtil.reduceToTopK(indexSearcher.getTaskExecutor().invokeAll(tasks), k);
+    }
+
+    private Map<Integer, Float> firstPassSearchLeaf(LeafReaderContext ctx, KNNWeight queryWeight) throws IOException {
+        Map<Integer, Float> leafDocScores = queryWeight.searchLeaf(ctx);
+        if (leafDocScores.isEmpty()) {
+            return leafDocScores;
+        }
+
+        Bits liveDocs = ctx.reader().getLiveDocs();
+        // TODO: This filters out the live docs from the knnWeight. Once knnWeight.searchLeaf returns results with
+        // deleted docs filtered out, this can be removed.
+        return leafDocScores.entrySet()
+            .stream()
+            .filter(entry -> liveDocs == null || liveDocs.get(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private TopDocs[] rescoreResults(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        List<Map<Integer, Float>> perLeafResults,
+        int k
+    ) throws IOException {
+        List<Callable<TopDocs>> topDocTasks = new ArrayList<>(leafReaderContexts.size());
+        for (int i = 0; i < perLeafResults.size(); i++) {
+            BitSet convertedBitSet = resultUtil.resultMapToMatchBitSet(perLeafResults.get(i));
```

Comment:
Any reason this is not inside the callable? Verify once but I think there is a loop inside this when creating bitset, throwing inside the callable will help with latencies

Reply:
I see that makes sense to have inside.